### PR TITLE
[Java] Package native dependencies into jar

### PR DESCRIPTION
**Pull request by raulchen**
## What do these changes do?

Package Java binary dependencies into the jar file, so the runtime won't depend on any relative paths. This change would make it easier to support maven and bazel simultaneously (#4284, cc @alegithub111 ).

Later, we can also build and distribute jars for mainstream platforms (just like the python wheels), so users don't need to compile the code by themselves.

### CODE REVIEW

**Comment by guoyuhong** [comment]
Could this PR solve the problem of building Arrow using bazel in https://github.com/ray-project/ray/pull/4284 ? So we can pre-build arrow and put all the binaries to a jar file like the python wheel and during building of ray, we need only to download the jar file, right?

**Comment by raulchen** [comment]
@guoyuhong yeah, we can do that.

File: java/runtime/src/main/java/org/ray/runtime/runner/RunManager.java [hunk 1]
```diff
@@ -76,8 +84,30 @@ private void createTempDirs() {
     FileUtil.mkDir(new File(rayConfig.objectStoreSocketName).getParentFile());
   }
 
+  /**
+   * Copy a file from resources to a temp dir, and return the file object.
+   */
+  private File getTempFile(String fileName) {
```
[the remark this answers is not in the document]
Comment (raulchen) [reply java/runtime/src/main/java/org/ray/runtime/runner/RunManager.java]: non-static should be fine. there's only one `RunManager` instance.

File: java/runtime/src/main/java/org/ray/runtime/RayNativeRuntime.java [hunk 2]
```diff
@@ -72,11 +77,22 @@ private void resetLibraryPath() {
 
   @Override
   public void start() throws Exception {
-    // Load native libraries.
     try {
+      // Reset library path at runtime.
       resetLibraryPath();
```
Comment (guoyuhong) [inline java/runtime/src/main/java/org/ray/runtime/RayNativeRuntime.java]: Do we still need this function to set `java.library.path`?

**Review by jovany-wang — APPROVED** [review]

**Review by guoyuhong — APPROVED** [review]
LGTM.

File: java/runtime/src/main/java/org/ray/runtime/RayNativeRuntime.java (hunk 2 above)
Comment (raulchen) [reply java/runtime/src/main/java/org/ray/runtime/RayNativeRuntime.java]: yeah, because users may pass in custom library path.

**Pull request merged**